### PR TITLE
fix data race in import unit(loader)

### PR DESCRIPTION
**Pull request by GregoryIan**
### What problem does this PR solve? 
find a data race in import unit(loader)
```
panic: send on closed channel

goroutine 44191 [running]:
github.com/pingcap/dm/loader.(*Worker).run.func1()
	/home/jenkins/agent/workspace/build_dm_master/go/src/github.com/pingcap/dm/loader/loader.go:155 +0x47b
created by github.com/pingcap/dm/loader.(*Worker).run
	/home/jenkins/agent/workspace/build_dm_master/go/src/github.com/pingcap/dm/loader/loader.go:175 +0x2ba
panic: send on closed channel
```

### What is changed and how it works?
* make sure the worker clean up all its goroutine when it exits
* It's very weird, dump unit start a `doJob` groutine for each dump files, not for each worker

### CODE REVIEW

File: loader/loader.go [hunk 1]
```diff
@@ -105,24 +105,30 @@ func NewWorker(loader *Loader, id int) (worker *Worker, err error) {
...
 func (w *Worker) run(ctx context.Context, fileJobQueue chan *fileJob, workerWg *sync.WaitGroup, runFatalChan chan *pb.ProcessError) {
 	atomic.StoreInt64(&w.closed, 0)
-	defer workerWg.Done()
 
 	newCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	defer func() {
+		cancel()
+		w.Close()
+		workerWg.Done()
```
[the remark this answers is not in the document]
Comment (GregoryIan) [reply loader/loader.go]: I don't change any logic

File: loader/loader.go [hunk 2]
```diff
@@ -105,24 +105,30 @@ func NewWorker(loader *Loader, id int) (worker *Worker, err error) {
 // Close closes worker
 func (w *Worker) Close() {
 	if !atomic.CompareAndSwapInt64(&w.closed, 0, 1) {
+		w.wg.Wait()
+		w.tctx.L().Info("already closed...")
```
[the remark this answers is not in the document]
Comment (GregoryIan) [reply loader/loader.go]: indeed

File: loader/loader.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply loader/loader.go]: I guess the problem may caused by the lack of `l.workerWg.Wait()` in `Loader.restoreData`.

[the remark this answers is not in the document]
Comment (GregoryIan) [reply loader/loader.go]: see code L1134

[the remark this answers is not in the document]
Comment (GregoryIan) [reply loader/loader.go]: another goroutine leak 🤔

**Comment by csuzhangxc** [comment]
@WangXiangUSTC PTAL

File: loader/loader.go [hunk 3]
```diff
@@ -104,32 +104,49 @@ func NewWorker(loader *Loader, id int) (worker *Worker, err error) {
...
-				w.tctx.L().Debug("execution goroutine exits")
+				w.tctx.L().Info("context canceled, execution goroutine exits")
 				return
 			case job, ok := <-w.jobQueue:
-				if !ok || job == nil {
+				if job == nil {
+					w.tctx.L().Info("jobs are finished, execution goroutine exits")
+					return
+				}
+				if !ok {
```
Comment (WangXiangUSTC) [inline loader/loader.go]: if `jobQueue` is closed, `job` is nil, and `ok` is false,  so it is better to put this statement before L144

File: tests/import_goroutine_leak/run.sh [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    check_port_offline $WORKER2_PORT 20
+
+    # dm-worker1 panics
+    err_cnt=`grep "panic" $WORK_DIR/worker1/log/stdout.log | wc -l`
+    if [ $err_cnt -ne 1 ]; then
+        echo "dm-worker1 doesn't panic, panic count ${err_cnt}"
+        exit 2
+    fi
+
+    echo "dm-workers paninc again, workers of import unit don't exit"
```
Comment (WangXiangUSTC) [inline tests/import_goroutine_leak/run.sh]:
```suggestion
    echo "dm-workers panic again, workers of import unit don't exit"
```

**Comment by WangXiangUSTC** [comment]
rest LGTM

**Comment by WangXiangUSTC** [comment]
LGTM

File: loader/loader.go [hunk 5]
```diff
@@ -104,32 +104,49 @@ func NewWorker(loader *Loader, id int) (worker *Worker, err error) {
...
 
 	ctctx := w.tctx.WithContext(newCtx)
 
 	doJob := func() {
-		defer w.wg.Done()
 		for {
 			select {
 			case <-newCtx.Done():
-				w.tctx.L().Debug("execution goroutine exits")
+				w.tctx.L().Info("context canceled, execution goroutine exits")
```
Comment (csuzhangxc) [inline loader/loader.go]: Will `Info` output too many logs?

File: loader/loader.go [hunk 6]
```diff
@@ -536,13 +571,24 @@ func (l *Loader) Restore(ctx context.Context) error {
...
+	failpoint.Inject("dontWaitWorkerExit", func(_ failpoint.Value) {
+		l.tctx.L().Info("", zap.String("failpoint", "dontWaitWorkerExit"))
+		failpoint.Return(nil)
+	})
+
+	// make sure all workers exit
+	l.closeFileJobQueue() // all data file dispatched, close it
+	l.workerWg.Wait()
+	if err == nil {
+		l.tctx.L().Info("all data files have been finished", zap.Duration("cost time", time.Since(begin)))
```
Comment (csuzhangxc) [inline loader/loader.go]: should we need to output duration for `context.Canceled`?

File: loader/loader.go (hunk 5 above)
Comment (GregoryIan) [reply loader/loader.go]: really not many

File: loader/loader.go (hunk 6 above)
Comment (GregoryIan) [reply loader/loader.go]: no need I think

File: loader/loader.go [hunk 7]
```diff
@@ -536,13 +571,24 @@ func (l *Loader) Restore(ctx context.Context) error {
 
 	go l.PrintStatus(ctx)
 
-	if err := l.restoreData(ctx); err != nil {
-		if errors.Cause(err) == context.Canceled {
-			return nil
-		}
+	begin := time.Now()
+	err = l.restoreData(ctx)
+	if err != nil && errors.Cause(err) != context.Canceled {
```
Comment (csuzhangxc) [inline loader/loader.go]: Is it needed
```
l.closeFileJobQueue()
l.workerWg.Wait()
```
before `return err`?

File: loader/loader.go [hunk 8]
```diff
@@ -192,12 +218,17 @@ func (w *Worker) restoreDataFile(ctx context.Context, filePath string, offset in
...
+		w.tctx.L().Info("", zap.String("failpoint", "dispatchError"))
+		failpoint.Return(errors.New("inject failpoint dispatchError"))
+	})
+
+	// dispatchSQL completed, send nil to make sure all dmls are applied to target database
 	// we don't want to close and re-make chan frequently
 	// but if we need to re-call w.run, we need re-make jobQueue chan
 	w.jobQueue <- nil
-
 	w.wg.Wait()
```
Comment (csuzhangxc) [inline loader/loader.go]: only `w.wg.Wait` in `Worker.Close` is enough now?

File: tests/import_goroutine_leak/run.sh [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    check_port_offline $WORKER1_PORT 20
+    check_port_offline $WORKER2_PORT 20
+
+    # dm-worker1 panics
+    err_cnt=`grep "panic" $WORK_DIR/worker1/log/stdout.log | wc -l`
+    if [ $err_cnt -ne 1 ]; then
+        echo "dm-worker1 doesn't panic, panic count ${err_cnt}"
+        exit 2
+    fi
```
Comment (csuzhangxc) [inline tests/import_goroutine_leak/run.sh]: Is it needed to check panic for worker2?

File: loader/loader.go (hunk 8 above)
Comment (GregoryIan) [reply loader/loader.go]: It means the file was finished here, the tricky point is that `w.wg` applied on both `file job` and `worker`

File: loader/loader.go (hunk 7 above)
Comment (GregoryIan) [reply loader/loader.go]: I thought it, but I don't change the origin logic, let it be

File: tests/import_goroutine_leak/run.sh (hunk 9 above)
Comment (GregoryIan) [reply tests/import_goroutine_leak/run.sh]: no sense for it, worker2 is same with worker1

Comment (GregoryIan) [reply tests/import_goroutine_leak/run.sh]: the framework need dm-worker2 must be there 😞

**Pull request merged**